pkg/controller/istiocontrolplane/istiocontrolplane_controller.go

```diff
@@ -109,8 +111,16 @@ func (r *ReconcileIstioControlPlane) Reconcile(request reconcile.Request) (recon
 		return reconcile.Result{}, err
 	}
 
-	deleted := instance.GetDeletionTimestamp() != nil
-	finalizers := instance.GetFinalizers()
+	// Workaroud for issue: https://github.com/istio/istio/issues/17883
+	// Using an unstructured object to get deletionTimestamp and finalizers fields.
+	u := &unstructured.Unstructured{}
```

Comment:
@morvencao 

I appreciate the work.  The operator should be using versioned client libraries for reading and writing IstioControlPlane's CR (as well as listening for changes, etc).  I think we can merge this as is.  The code base is in a need of overhaul to properly provide a clientset/informers/listeners that are compatible with `controller-runtime`.  I've been banging on it for about 6 hours, and have something that nearly almost compiles..., but it is very disruptive and I am not keen to merge this work prior to the release of 1.4.

I will however post my work a a nonmergeable PR so folks can see what I am talking about in this comment :)

We should probably do a more thorough audit of the operator code base for these problems.

Cheers
-steve

Reply:
@sdake Understood.
As a workaround, I really hope this can be merged so that we can continue to test more cases and also make e2e tests(coming soon I think) work as expected, the original issue https://github.com/istio/istio/issues/17883 is a blocker for us to proceed.

Maybe I should keep the original github issue https://github.com/istio/istio/issues/17883 in open state until the API marshal/unmarshal issue is completely resolved.

/cc @ostromart

Reply:
yes please keep the issue open.  I think we need to make an exit from gogo - should simplify the implementation.